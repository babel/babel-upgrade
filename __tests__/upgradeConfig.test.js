const path = require('path');
const upgradeConfig = require('../src/upgradeConfig');
const babelrcFixture = require('../fixtures/babelrc');
const optionParsingFixture = require('../fixtures/option-parsing');
const dupesProspectFixture = require('../fixtures/dupes-prospect');
const { readBabelRC } = require('../src');
const JSON5_PATH = path.resolve(__dirname, '../fixtures/babelrc.json5');

test('packages', () => {
  expect(upgradeConfig(babelrcFixture)).toMatchSnapshot();
});

test('new plugins with array', () => {
  expect(upgradeConfig({
    plugins: [
      'babel-plugin-syntax-async-generators',
      'babel-plugin-syntax-export-extensions',
      ["babel-plugin-transform-es2015-arrow-functions", { "spec": true }]
    ]
  })).toMatchSnapshot();
})

test('package that is removed', () => {
  expect(upgradeConfig({
    plugins: [
      'transform-function-bind',
      'check-es2015-constants',
      'transform-es2015-spread',
      'instanbul'
    ]
  })).toMatchSnapshot();
});

test('rename community packages', () => {
  expect(upgradeConfig({
    plugins: [
      'lodash'
    ],
    env: {
      development: {
        plugins: [
          'lodash'
        ]
      }
    }
  })).toMatchSnapshot();
});

test('packages (json5)', async () => {
  const json5Data = await readBabelRC(JSON5_PATH);
  expect(upgradeConfig(json5Data)).toMatchSnapshot();
});

test('convert comma separated presets/plugins into an array', () => {
  expect(upgradeConfig(optionParsingFixture)).toMatchSnapshot();
});

test('adds flow preset if hasFlow option passed', () => {
  const config = {
    "presets": ["react"],
  };

  expect(upgradeConfig(config, { hasFlow: true })).toMatchSnapshot();
});

test('does not add flow preset if hasFlow option is not passed', () => {
  const config = {
    "presets": ["react"],
  };

  expect(upgradeConfig(config, { hasFlow: false })).toMatchSnapshot();
});

test('does not another flow preset if already present and hasFlow option passed', () => {
  const config = {
    "presets": [
      "@babel/preset-react",
      "@babel/preset-flow",
    ],
  };

  expect(upgradeConfig(config, { hasFlow: true })).toMatchSnapshot();
});

test('handles @babel prefix in plugins', () => {
  const config = {
    "plugins": [
      '@babel/plugin-transform-async-generator-functions',
      '@babel/plugin-transform-class-properties',
      '@babel/plugin-transform-do-expressions'
    ],
  };

  expect(upgradeConfig(config)).toMatchSnapshot();
});

test("adds legacy option to decorators", () => {
  const config = {
    "plugins": [
      "transform-decorators",
      "@babel/plugin-syntax-decorators"
    ],
  };

  expect(upgradeConfig(config)).toMatchSnapshot();
});

test("adds proposal option to pipeline", () => {
  const config = {
    "plugins": [
      "transform-pipeline-operator",
      "@babel/plugin-syntax-pipeline-operator"
    ],
  };

  expect(upgradeConfig(config)).toMatchSnapshot();
});

test("replaces stage presets", () => {
  const config = {
    "presets": [
      "stage-1"
    ]
  };

  expect(upgradeConfig(config)).toMatchSnapshot();
});

test("prevent dupe plugins", () => {
  expect(upgradeConfig(dupesProspectFixture)).toMatchSnapshot();
});
