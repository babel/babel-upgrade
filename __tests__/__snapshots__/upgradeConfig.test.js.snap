// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds flow preset if hasFlow option passed 1`] = `
Object {
  "presets": Array [
    "@babel/preset-react",
    "@babel/preset-flow",
  ],
}
`;

exports[`adds legacy option to decorators 1`] = `
Object {
  "plugins": Array [
    Array [
      "@babel/plugin-proposal-decorators",
      Object {
        "legacy": true,
      },
    ],
    Array [
      "@babel/plugin-syntax-decorators",
      Object {
        "legacy": true,
      },
    ],
  ],
}
`;

exports[`adds proposal option to pipeline 1`] = `
Object {
  "plugins": Array [
    Array [
      "@babel/plugin-proposal-pipeline-operator",
      Object {
        "proposal": "minimal",
      },
    ],
    Array [
      "@babel/plugin-syntax-pipeline-operator",
      Object {
        "proposal": "minimal",
      },
    ],
  ],
}
`;

exports[`convert comma separated presets/plugins into an array 1`] = `
Object {
  "env": Object {
    "development": Object {
      "plugins": Array [
        "babel-plugin-istanbul",
        "@babel/plugin-transform-react-jsx-source",
        "@babel/plugin-transform-react-inline-elements",
      ],
    },
  },
  "plugins": Array [
    "@babel/plugin-proposal-function-bind",
    "@babel/plugin-proposal-class-properties",
  ],
  "presets": Array [
    "@babel/preset-env",
    "@babel/preset-react",
  ],
}
`;

exports[`does not add flow preset if hasFlow option is not passed 1`] = `
Object {
  "presets": Array [
    "@babel/preset-react",
  ],
}
`;

exports[`does not another flow preset if already present and hasFlow option passed 1`] = `
Object {
  "presets": Array [
    "@babel/preset-react",
    "@babel/preset-flow",
  ],
}
`;

exports[`handles @babel prefix in plugins 1`] = `
Object {
  "plugins": Array [
    "@babel/plugin-proposal-async-generator-functions",
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-do-expressions",
  ],
}
`;

exports[`new plugins with array 1`] = `
Object {
  "plugins": Array [
    "@babel/plugin-syntax-async-generators",
    "@babel/plugin-syntax-export-default-from",
    "@babel/plugin-syntax-export-namespace-from",
    Array [
      "@babel/plugin-transform-arrow-functions",
      Object {
        "spec": true,
      },
    ],
  ],
}
`;

exports[`package that is removed 1`] = `
Object {
  "plugins": Array [
    "@babel/plugin-proposal-function-bind",
    "@babel/plugin-transform-spread",
    "instanbul",
  ],
}
`;

exports[`packages (json5) 1`] = `
Object {
  "env": Object {
    "development": Object {
      "plugins": Array [
        "babel-plugin-istanbul",
        "@babel/plugin-transform-react-jsx-source",
        "@babel/plugin-transform-react-jsx-self",
      ],
    },
    "production": Object {
      "plugins": Array [
        "@babel/plugin-transform-react-constant-elements",
        "@babel/plugin-transform-react-inline-elements",
      ],
    },
  },
  "plugins": Array [
    "@babel/plugin-proposal-function-bind",
    Array [
      "@babel/plugin-proposal-object-rest-spread",
      Object {
        "useBuiltIns": true,
      },
    ],
    "@babel/plugin-proposal-class-properties",
  ],
  "presets": Array [
    "@babel/preset-env",
    Array [
      "@babel/preset-env",
      Object {
        "modules": false,
      },
    ],
    "@babel/preset-react",
  ],
}
`;

exports[`packages 1`] = `
Object {
  "env": Object {
    "development": Object {
      "plugins": Array [
        "babel-plugin-istanbul",
        "@babel/plugin-transform-react-jsx-source",
        "@babel/plugin-transform-react-jsx-self",
      ],
    },
    "production": Object {
      "plugins": Array [
        "@babel/plugin-transform-react-constant-elements",
        "@babel/plugin-transform-react-inline-elements",
      ],
    },
  },
  "plugins": Array [
    "@babel/plugin-proposal-function-bind",
    Array [
      "@babel/plugin-proposal-object-rest-spread",
      Object {
        "useBuiltIns": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-decorators",
      Object {
        "legacy": true,
      },
    ],
    "@babel/plugin-proposal-class-properties",
  ],
  "presets": Array [
    "@babel/preset-env",
    Array [
      "@babel/preset-env",
      Object {
        "modules": false,
      },
    ],
    "@babel/preset-react",
  ],
}
`;

exports[`prevent dupe plugins 1`] = `
Object {
  "env": Object {
    "development": Object {
      "plugins": Array [
        "react-hot-loader/babel",
      ],
    },
    "production": Object {
      "plugins": Array [
        "transform-react-remove-prop-types",
      ],
    },
    "test": Object {
      "plugins": Array [
        "@babel/plugin-transform-runtime",
        Array [
          "@babel/plugin-proposal-decorators",
          Object {
            "legacy": true,
          },
        ],
        "@babel/plugin-transform-react-display-name",
        "@babel/plugin-proposal-class-properties",
        Array [
          "import",
          Object {
            "libraryName": "antd",
            "style": "css",
          },
        ],
      ],
      "presets": Array [
        Array [
          "@babel/preset-env",
          Object {
            "loose": true,
            "modules": false,
            "targets": Object {
              "browsers": Array [
                "> 0.25%",
              ],
            },
            "useBuiltIns": "usage",
          },
        ],
        "@babel/preset-react",
      ],
    },
  },
  "plugins": Array [
    Array [
      "@babel/plugin-proposal-decorators",
      Object {
        "legacy": true,
      },
    ],
    "@babel/plugin-transform-react-display-name",
    "@babel/plugin-proposal-class-properties",
    Array [
      "import",
      Object {
        "libraryName": "antd",
        "style": "css",
      },
    ],
    "@babel/plugin-syntax-dynamic-import",
    "@babel/plugin-syntax-import-meta",
    "@babel/plugin-proposal-json-strings",
    "@babel/plugin-proposal-function-sent",
    "@babel/plugin-proposal-export-namespace-from",
    "@babel/plugin-proposal-numeric-separator",
    "@babel/plugin-proposal-throw-expressions",
    "@babel/plugin-proposal-export-default-from",
    "@babel/plugin-proposal-logical-assignment-operators",
    "@babel/plugin-proposal-optional-chaining",
    Array [
      "@babel/plugin-proposal-pipeline-operator",
      Object {
        "proposal": "minimal",
      },
    ],
    "@babel/plugin-proposal-nullish-coalescing-operator",
    "@babel/plugin-proposal-do-expressions",
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "loose": true,
        "modules": false,
        "targets": Object {
          "browsers": Array [
            "> 0.25%",
          ],
        },
        "useBuiltIns": "usage",
      },
    ],
    "@babel/preset-react",
    "@babel/preset-flow",
  ],
}
`;

exports[`rename community packages 1`] = `
Object {
  "env": Object {
    "development": Object {
      "plugins": Array [
        "lodash",
      ],
    },
  },
  "plugins": Array [
    "lodash",
  ],
}
`;

exports[`replaces stage presets 1`] = `
Object {
  "plugins": Array [
    "@babel/plugin-syntax-dynamic-import",
    "@babel/plugin-syntax-import-meta",
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-json-strings",
    Array [
      "@babel/plugin-proposal-decorators",
      Object {
        "legacy": true,
      },
    ],
    "@babel/plugin-proposal-function-sent",
    "@babel/plugin-proposal-export-namespace-from",
    "@babel/plugin-proposal-numeric-separator",
    "@babel/plugin-proposal-throw-expressions",
    "@babel/plugin-proposal-export-default-from",
    "@babel/plugin-proposal-logical-assignment-operators",
    "@babel/plugin-proposal-optional-chaining",
    Array [
      "@babel/plugin-proposal-pipeline-operator",
      Object {
        "proposal": "minimal",
      },
    ],
    "@babel/plugin-proposal-nullish-coalescing-operator",
    "@babel/plugin-proposal-do-expressions",
  ],
  "presets": Array [],
}
`;
